## 서론

`Flow`는 한 쪽에서는 값을 요청하고, 반대쪽에서는 요청을 본 후에 값을 준비하는 파이프라고 할 수 있다.

`Flow`에서 예외가 발생하거나 complete되었다면 해당 정보는 부모에게 전파되며, 중간에 중간 단계를 실행한다.

`Flow는 값, 예외, 또는 특정 이벤트를 Listen할 수 있다.

이를 위해 `onEach()`, `onStart()`, `onCompletion()`, `onEmpty()`, `catch()`와 같은 메소드를 지원한다.

## onEach

들어오는 값에 대한 동작을 정의하기 위해 `onEach()`를 사용할 수 있다.

``` kotlin
suspend fun main() {
    flowOf(1, 2, 3, 4)
		.onEach { print(it) }
        .collect() // 1234
}
```

람다 표현식은 suspend로, 각각의 Element들은 순서대로 처리된다.

즉, **`onEach()`에서 suspend된다면 다음 Element는 처리되지 않는다.**

``` kotlin
suspend fun main() {
    flowOf(1, 2)
        .onEach { delay(1000) }
        .collect { println(it) }
}

// 1초 뒤
1
// 1초 뒤
2
```

## onStart

`onStart()` 함수는 **`Flow`가 시작되자마자 호출되는 Listener**이다.

즉, **종단 연산자가 실행 된 후 즉시 실행**된다.

`onStart()`는 첫 번째 Element를 기다리지 않고, 정말 종단 연산자가 호출되자마자 실행된다.

``` kotlin
suspend fun main() {
    flowOf(1, 2)
        .onEach { delay(1000) }
        .onStart { println("Before") }
        .collect { println(it) }
}

Before
// 1초 뒤
1
// 1초 뒤
2
```

`onStart()`에서는 Element를 emit할 수 있다.

> `onCompletion()`, `onEmpty()`, `catch()`에서도 마찬가지이다.

``` kotlin
suspend fun main() {
    flowOf(1, 2)
        .onEach { delay(1000) }
        .onStart { emit(0) }
        .collect { println(it) }
}

0
// 1초 뒤
1
// 1초 뒤
2
```

## onCompletion

`Flow`가 complete되는데에는 몇 가지 방법이 있는데, 가장 많이 사용되는 방법이 `Flow`가 종료되는 것이다.

마지막 Element가 전송된 후를 의미하는데, **`catch()`에서 걸리지지 않은 예외가 발생하거나, coroutine이 cancel되었을 때도 호출된다.**

`onCompletion()` 메소드를 통해 이 모든 경우를 포함하는 Listener를 등록할 수 있다.

``` kotlin
suspend fun main() = coroutineScope {
    flowOf(1, 2)
        .onEach { delay(1000) }
        .onCompletion { println("Completed") }
        .collect { println(it) }
}

// 1초 뒤
1
// 1초 뒤
2
Completed

suspend fun main() = coroutineScope {
    val job = launch {
        flowOf(1, 2)
            .onEach { delay(1000) }
            .onCompletion { println("Completed") }
            .collect { println(it) }
    }
    delay(1100)
    job.cancel()
}

// 1초 뒤
1
// 0.1초 뒤
Completed
```

Android에서는 주로 `onStart()`에서 progressbar를 띄우고, `onCompletion()`에서 다시 이걸 숨기는데에 사용한다.

``` kotlin
fun updateNews() {
    scope.launch {
        newsFlow()
        	.onStart { showProgressBar() }
        	.onCompletion { hideProgressBar() }
        	.collect { view.showNews(it) }
    }
}
```

## onEmpty

`Flow`는 어떤 값도 없이 바로 종료될수도 있는데, 특정 상황에선 예기치 못한 상황일수도 있다.

이런 경우를 위해 **`Flow`가 비었을 때 실행되는 `onEmpty()`함수가 존재한다.**

``` kotlin
suspend fun main() = coroutineScope {
    flow<List<Int>> { delay(1000) }
        .onEmpty { emit(emptyList()) }
        .collect { println(it) }
}

// 1초 뒤
[]
```

`Flow`가 비었을 때, default 값을 주고싶을 때도 사용할 수 있다.

## catch

`Flow`를 생성하거나, 처리하는 등 어디서나 예외는 발생할 수 있다.

Exception은 내려가며 각각의 처리를 순차적으로 멈춘다,

하지만, 이러한 **예외를 따로 처리하고 싶을수도 있기 때문에 `catch()`함수를 지원한다.**

이 Listener는 Exception을 파라미터로 받아 복원 작업을 할 수 있다.

``` kotlin
class MyError : Throwable("My error")

val flow = flow {
    emit(1)
    emit(2)
    throw MyError()
}

suspend fun main() {
    flow.onEach { println("Got $it") }
    	.catch { println("Caught $it") }
    	.collect { println("Collected $it") }
}

Got 1
Collected 1
Got 2
Collected 2
Caught MyError: My error
```

> 위에서 볼 수 있다싶이 `onEach()`는 예외에 대해서는 실행되지 않는다.

`catch()`는 예외가 퍼지는걸 막아준다.

즉 **`catch()`에서 새로 값을 emit하면 처리해줄 수 있다.**

``` kotlin
val flow = flow {
    emit("Message1")
    throw MyError()
}

suspend fun main() {
    flow.catch { emit("Error") }
    	.collect { println("Collected $it") }
}

Collected Message1
Collected Error
```

`catch()`는 **upstream에서 발생한 예외만 처리해 준다.**

즉, 자신의 downstream에서 발생한 예외는 처리해 주지 않기 때문에 다음 코드에서는 예외가 잡히지 않는다.

``` kotlin
suspend fun main() {
    flowOf("Messsage1")
    	.catch { emit("Error") }
    	.onEach { throw Error(it) }
    	.collect { println("Collected $it") }
}
```

Android는 `Flow`에서 발생한 예외를 보여줄 때 사용한다.

``` kotlin
fun updateNews() {
    scope.launch {
        newsFlow()
            .catch { view.handleError(it) }
            .onStart { showProgressBar() }
            .onCompletion { hideProgressBar() }
            .collect { view.showNews(it) }
    }
}
```

또한 예외가 발생했을 때 기본 값을 보여주고 싶을 때도 사용할 수 있다.

``` kotlin
fun updateNews() {
    scope.launch {
        newsFlow()
            .catch {
                view.handleError(it)
                emit(emptyList())
            }
            .onStart { showProgressBar() }
        	.onCompletion { hideProgressBar() }
            .collect { view.showNews(it) }
    }
}
```

## Uncaught Exception

`Flow`에서 잡히지 않은 예외는 **즉시 `Flow`를 종료하고, `collect()`에서 예외를 다시 던져준다.**

일반적인 suspend함수와 같고, `coroutineScope`도 똑같이 동작한다.

그러므로 일반적인 `try - catch`에서 처리할 수 있다.

``` kotlin
val flow = flow {
    emit("Message1")
    throw MyError()
}

suspend fun main() {
    try {
        flow.collect { println("Collected $it") }
    } catch (e: MyError) {
        println("Caught")
    }
}

Collected Message1
Caught
```

**`catch()`는 종단 연산자에서 발생하는 예외를 잡아주지 못한다.**

따라서 만약 `collect()`에서 예외가 발생할 때에는 `catch()`로 잡을순 없고 `try - catch`구문으로 막아야 한다.

``` kotlin
val flow = flow {
    emit("Message1")
    emit("Message2")
}

suspend fun main() {
    flow.onStart { println("Before") }
    	.catch { println("Caught $it") }
    	.collect { throw MyError() }
}

Before
Exception in thread "..." MyError: My error
```

그래서 `collect()`에 연산자를 두기보다는 `catch()`보다 위에 `onEach()`를 둬서 예외를 처리하는게 일반적이다.

`Flow` 바깥쪽으로 예외를 던지지 않고, `catch()`에서 처리함을 보장해 준다.

``` kotlin
val flow = flow {
    emit("Message1")
    emit("Message2")
}

suspend fun main() {
    flow.onStart { println("Before") }
    	.onEach { throw MyError() }
    	.catch { println("Caught $it") }
    	.collect()
}

Before
Caught MyError: My error
```

## flowOn

`Flow`의 연산자들은 람다 표현식을 파라미터로 받고, `Flow`의 builder들은 기본적으로 모두 suspend된다.

suspend 함수들은 Context가 있어야 하고, 부모와 관계를 가져야 한다.

이 함수들이 Context를 어디서 얻을지, 그리고 어떻게 바꿀지 궁금할 수도 있다.

기본적으로 **`Flow`는 `collect()`가 호출된 곳의 Context를 가진다.**

``` kotlin
fun usersFlow(): Flow<String> = flow {
    repeat(2) {
        val ctx = currentCoroutineContext()
        val name = ctx[CoroutineName]?.name
        emit("User$it in $name")
    }
}

suspend fun main() {
    val users = usersFlow()
    withContext(CoroutineName("Name1")) {
        users.collect { println(it) }
    }
    withContext(CoroutineName("Name2")) {
        users.collect { println(it) }
    }
}

User0 in Name1
User1 in Name1
User0 in Name2
User1 in Name2
```

종단 연산자들은 upstream으로부터 Element를 요청한다.

그와 동시에, **Coroutine Context가 전파된다.**

이걸 따로 커스텀 해주고 싶다면 `flowOn()`을 사용할 수 있다.

``` kotlin
suspend fun present(place: String, message: String) {
    val ctx = coroutineContext
    val name = ctx[CoroutineName]?.name
    println("[$name] $message on $place")
}

fun messagesFlow(): Flow<String> = flow {
    present("flow builder", "Message")
    emit("Message")
}

suspend fun main() {
    val users = messagesFlow()
    
    withContext(CoroutineName("Name1")) {
        users
            .flowOn(CoroutineName("Name3"))
            .onEach { present("onEach", it) }
            .flowOn(CoroutineName("Name2"))
            .collect { present("collect", it) }
    }
}

[Name3] Message on flow builder
[Name2] Message on onEach
[Name1] Message on collect
```

**`flowOn()`은 upstream의 함수에 대해서만 동작한다.**

위에서 볼 수 있다싶이, 중간 연산자 밑에 `flowOn()`을 붙임으로써 해당 연산자의 Context를 설정해 줬다ㅣ.

## launchIn

`collect()`는 `Flow`가 종단 상태가 될 때까지 suspend 시키는 연산자이다.

보통 `launch` builder로 묶어주곤 하는데, 그 이유라고 할 수 있다.

이걸 좀 더 편하게 사용하기 위한 `launchIn()`함수가 있는데, **새로운 Coroutine(파라미터로 넘긴 Scope)에서 연산자를 실행시킨다.**

``` kotlin
fun <T> Flow<T>.launchIn(scope: CoroutineScope): Job =
	scope.launch { collect() }
```

`launchIn()`은 **여러 Coroutine에서 `Flow`를 처리하고 싶을 때 사용할 수 있다.**

``` kotlin
suspend fun main(): Unit = coroutineScope {
    flowOf("User1", "User2")
    	.onStart { println("Users:") }
    	.onEach { println(it) }
    	.launchIn(this)
}

// Users:
// User1
// User2
```

